File: 2023/kotlin/src/Day20.kt
private typealias Id = String

private data class Pulse(val from: Id, val to: Id, val isHigh: Boolean)

private abstract class Module(open val id: Id, open val incoming: List<Id>, open val outgoing: List<Id>) {
    abstract fun handle(pulse: Pulse): List<Pulse>

    companion object {
        fun of(input: String): Pair<Broadcast, Map<String, Module>> {
            val mappings = input.lines().map {
                val (id, outgoing) = it.split(" -> ")
                id to outgoing.split(", ")
            }

            val broadcaster = Broadcast(mappings.first { it.first == "broadcaster" }.second)

            val modules = mappings.filter { it.first != "broadcaster" }.associate { mapping ->
                val (typeAndId, outgoing) = mapping
                val type = typeAndId.first()
                val id = typeAndId.drop(1)

                val incoming = mappings.mapNotNull { if (it.second.contains(id)) it.first.drop(1) else null }

                val module = when (type) {
                    '%' -> FlipFlop(id, incoming, outgoing)
                    else -> Conjunction(id, incoming, outgoing)
                }

                id to module
            }

            return broadcaster to modules
        }
    }
}

private data class FlipFlop(override val id: String, override val incoming: List<String>, override val outgoing: List<String>): Module(id, incoming, outgoing) {
    private var wasPreviousPulseHigh = false

    override fun handle(pulse: Pulse): List<Pulse> {
        if (pulse.isHigh) return listOf()
        wasPreviousPulseHigh = !wasPreviousPulseHigh
        return outgoing.map { Pulse(id, it, wasPreviousPulseHigh) }
    }
}

private data class Conjunction(override val id: String, override val incoming: List<String>, override val outgoing: List<String>): Module(id, incoming, outgoing) {
    private var history: MutableMap<String, Boolean> = incoming.associateWith { false }.toMutableMap()

    override fun handle(pulse: Pulse): List<Pulse> {
        history[pulse.from] = pulse.isHigh
        return outgoing.map { outgoingId -> Pulse(id, outgoingId, !history.values.all { it }) }
    }
}

private data class Broadcast(val outgoing: List<String>) {
    fun press() = outgoing.map { Pulse("broadcaster", it, false) }
}

private fun part1(input: String): Long {
    val (broadcaster, modules) = Module.of(input)

    var lowPulses = 0L
    var highPulses = 0L
    val queue = ArrayDeque<Pulse>()

    repeat(1000) {
        lowPulses++
        queue.addAll(broadcaster.press())

        while (queue.isNotEmpty()) {
            val pulse = queue.removeFirst()
            if (pulse.isHigh) highPulses++ else lowPulses++
            modules[pulse.to]?.let { module -> queue.addAll(module.handle(pulse)) }
        }
    }

    return lowPulses * highPulses
}

private fun part2(input: String): Long {
    val (broadcaster, modules) = Module.of(input)

    var presses = 0L
    val queue = ArrayDeque<Pulse>()

    val lastModuleId = modules.values.first { it.outgoing.contains("rx") }.id
    val connectionsToLastModule = modules.values.mapNotNull { if (it.outgoing.contains(lastModuleId)) it.id else null }
    val pressesToHighPulse = connectionsToLastModule.associateWith { 0L }.toMutableMap()

    while (pressesToHighPulse.any { it.value == 0L }) {
        presses++
        queue.addAll(broadcaster.press())

        while (queue.isNotEmpty()) {
            val pulse = queue.removeFirst()
            if (pulse.to == lastModuleId && pulse.isHigh) pressesToHighPulse[pulse.from] = presses
            modules[pulse.to]?.let { module -> queue.addAll(module.handle(pulse)) }
        }
    }

    return pressesToHighPulse.values.reduce(::lcm)
}

fun main() {
    val input = readInput("Day20")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day20.swift
struct Day20: AdventDay {
  var input: String

  struct Pulse: CustomStringConvertible {
    let from: String, to: String, isHigh: Bool

    var description: String {
      "\(from) -\(isHigh)-> \(to)"
    }
  }

  class Module {
    let id: String, incoming: [String], outgoing: [String]

    init(id: String, incoming: [String], outgoing: [String]) {
      self.id = id
      self.incoming = incoming
      self.outgoing = outgoing
    }

    func handle(pulse: Pulse) -> [Pulse] {
      []
    }

    static func build(from input: String) -> (Broadcast, [String: Module]) {
      let mappings = input.lines().map { line in
        let parts = line.components(separatedBy: " -> ")
        let outgoing = parts[1].components(separatedBy: ", ")
        return (parts[0], outgoing)
      }

      let broadcaster = Broadcast(outgoing: mappings.first(where: { $0.0 == "broadcaster" })!.1)

      let modules = mappings.filter { $0.0 != "broadcaster" }.map { mapping in
        let (typeAndId, outgoing) = mapping
        let type = typeAndId.first!
        let id = String(typeAndId.dropFirst(1))

        let incoming = mappings.compactMap { mapping in
          let (typeAndId, outgoing) = mapping
          return outgoing.contains(id) ? String(typeAndId.dropFirst(1)) : nil
        }

        let module =
          type == "%"
          ? FlipFlop(id: id, incoming: incoming, outgoing: outgoing)
          : Conjunction(id: id, incoming: incoming, outgoing: outgoing)

        return (id, module)
      }

      return (broadcaster, Dictionary(uniqueKeysWithValues: modules))
    }
  }

  class FlipFlop: Module {
    private var wasPreviousPulseHigh = false

    override func handle(pulse: Pulse) -> [Pulse] {
      if pulse.isHigh { return [] }
      wasPreviousPulseHigh = !wasPreviousPulseHigh
      return outgoing.map { Pulse(from: id, to: $0, isHigh: wasPreviousPulseHigh) }
    }
  }

  class Conjunction: Module {
    private var history: [String: Bool]

    override init(id: String, incoming: [String], outgoing: [String]) {
      history = Dictionary(uniqueKeysWithValues: incoming.map { ($0, false) })
      super.init(id: id, incoming: incoming, outgoing: outgoing)
    }

    override func handle(pulse: Pulse) -> [Pulse] {
      history[pulse.from] = pulse.isHigh
      return outgoing.map { Pulse(from: id, to: $0, isHigh: !history.values.allSatisfy { $0 }) }
    }
  }

  struct Broadcast {
    let outgoing: [String]

    func press() -> [Pulse] {
      outgoing.map { Pulse(from: "broadcaster", to: $0, isHigh: false) }
    }
  }

  func part1() -> Any {
    let (broadcaster, modules) = Module.build(from: input)

    var lowPulses = 0
    var highPulses = 0
    var queue = [Pulse]()

    for _ in 0..<1000 {
      lowPulses += 1
      queue.append(contentsOf: broadcaster.press())

      while !queue.isEmpty {
        let pulse = queue.removeFirst()
        if pulse.isHigh { highPulses += 1 } else { lowPulses += 1 }
        if let module = modules[pulse.to] {
          queue.append(contentsOf: module.handle(pulse: pulse))
        }
      }
    }

    return lowPulses * highPulses
  }

  func part2() -> Any {
    let (broadcaster, modules) = Module.build(from: input)

    var presses = 0
    var queue = [Pulse]()

    let lastModuleId = modules.values.first(where: { $0.outgoing.contains("rx") })!.id
    let connectionsToLastModule = modules.values.compactMap {
      $0.outgoing.contains(lastModuleId) ? $0.id : nil
    }
    var pressesToHighPulse = Dictionary(
      uniqueKeysWithValues: connectionsToLastModule.map { ($0, 0) })

    while pressesToHighPulse.contains(where: { $0.value == 0 }) {
      presses += 1
      queue.append(contentsOf: broadcaster.press())

      while !queue.isEmpty {
        let pulse = queue.removeFirst()
        if pulse.to == lastModuleId && pulse.isHigh { pressesToHighPulse[pulse.from] = presses }
        if let module = modules[pulse.to] {
          queue.append(contentsOf: module.handle(pulse: pulse))
        }
      }
    }

    return pressesToHighPulse.values.reduce(1, *)
  }
}
